### PR TITLE
[FLINK-785] Chained AllReduce / AllGroupReduceCombine

### PR DESCRIPTION
**Pull request by zentol**
This a a preliminary PR to see whether I'm on the right track.

I'm wondering whether this would be everything needed to add a Chained AllReduce, before i continue with this issue. I tried it out and it appears to work, but wanted to make sure.

### CODE REVIEW

**Comment by fhueske** [comment]
Looks good in general. 
You need to make sure though, that you obey the execution object re-usage settings.
That basically means you need to pay attention to the objects that you give to and receive from the user code and possibly copy them to new objects (see FLINK-1285 and @aljoscha's commits for details).

**Comment by zentol** [comment]
there's something funky going on with the tests here.

i got 2 failing tests in ObjectReuseITCase:
```
ObjectReuseITCase>JavaProgramTestBase.testJobWithoutObjectReuse:168->postSubmit:68->TestBaseUtils.compareResultsByLinesInMemory:223->TestBaseUtils.compareResultsByLinesInMemory:238 arrays first differed at element [0]; expected:<a,[10]0> but was:<a,[6]0>

ObjectReuseITCase>JavaProgramTestBase.testJobWithObjectReuse:120->postSubmit:68->TestBaseUtils.compareResultsByLinesInMemory:223->TestBaseUtils.compareResultsByLinesInMemory:238 arrays first differed at element [0]; expected:<a,[10]0> but was:<a,[6]0>
```
These two tests verify the wrong behaviour that occurs when object reuse is enabled but not accounted for. i thought this was generally treated as undefined behaviour, why are there tests for that?

the other 2 tests fail with NullPointerException when accessing the expected result.
```
ClosureCleanerITCase.after:52->TestBaseUtils.compareResultsByLinesInMemory:223->TestBaseUtils.compareResultsByLinesInMemory:234 » NullPointer
ClosureCleanerITCase.after:52->TestBaseUtils.compareResultsByLinesInMemory:223->TestBaseUtils.compareResultsByLinesInMemory:234 » NullPointer
```
i can't figure out why this occurs.

**Comment by fhueske** [comment]
Hi @zentol,

I had a look at your PR and the failing test case. The test was in place to ensure that objects are reused. For that the functions were implemented in a way that they will produce wrong results if objects are reused (therefore the unexpected result of 100 instead of 60). The problem was caused by the CollectionInputFormat which sends out different objects (instead of reusing the same object). When I changed that behavior and added some fixes to the ChainedAllReduceDriver, all tests pass.
I will open a PR against your branch.

**Comment by fhueske** [comment]
Oh, just saw that you updated your PR.
Won't open a PR. You can have a look at my branch here: https://github.com/fhueske/flink/tree/chained_all_reduce

**Comment by zentol** [comment]
sorry for not making it clear that i forced pushed some changes, forgot making a separate branch.

the changes you made to the driver are already in, sorry for wasting your time there.
i will include the changes to the test case.

The ChainedAllReduceDriver currently does not cause any problems.
I've added a ChainedAllGroupReduceCombineDriver aswell, which fails for one test case
```
ReplicatingDataSourceITCase.after:69->TestBaseUtils.compareResultsByLinesInMemory:223->TestBaseUtils.compareResultsByLinesInMemory:238 arrays first differed at element [0]; expected:<([50050]0)> but was:<([37525]0)>
```
may be related to FLINK-1521, once the newest test is complete I'll hopefully have a definite answer. when i force objectReuse in the global reduce all tests pass.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllGroupReduceCombineDriver.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	public String getTaskName() {
+		return this.taskName;
+	}
+
+	// --------------------------------------------------------------------------------------------
+	@Override
+	public void collect(T record) {
+		try {
+			values.add(objectReuseEnabled ? record : serializer.copy(record));
+			if (values.size() > 10000) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllGroupReduceCombineDriver.java]: is there a more reasonable value for this? (basically anything whose reasoning goes beyond "i felt like using 10000")

**Comment by StephanEwen** [comment]
I think for the combiner, we cannot do this. The way it is done here easily blows up the memory, by collecting so many records. Bear in mind that a record is not necessarily small, in the ALS case, they were 500MB large. Times 10k, that is 5TB of buffered records. The non-chaining combiner variant makes sure that only two objects are ever around by passing records/fragments through the buffers in the local channels.

**Comment by StephanEwen** [comment]
The chained reducer variant makes more sense. The question is, what is cheaper - serialized exchange through the in-mem-channel, or the copying. My gut feeling is that the copy is a bit better, so it may be worth it.

**Comment by zentol** [comment]
the SynchronouesChainedCombineDriver stores multiple records as well doesn't it? except up to a fixed size instead of # of elements. could we not do the same here?

**Comment by StephanEwen** [comment]
The synchronous combine driver stores them in the sort buffer in serialized fashion, which is save.

**Comment by zentol** [comment]
i know, so could we not do the same, except skipping the sorting part.

**Comment by StephanEwen** [comment]
Then it is worse then the chained variant, because you save nothing and need buffer space. The chained variant forwards records/fragments as they are produces.

**Comment by StephanEwen** [comment]
I think we should use the chained reducer from this pull request and drop the chained combiner change.

**Comment by zentol** [comment]
okay, I will the remove the chained combiner.

I'm curious though, the issue description specifically mentions the AllGroupReduce greatly benefiting from a chained version. did (or do) you have a specific approach in mind?

**Pull request closed without merging**